txnsync/peer.go

```diff
@@ -0,0 +1,809 @@
+}
+
+// getLocalRequestParams returns the local requests params
+func (p *Peer) getLocalRequestParams() (offset, modulator byte) {
+	return p.localTransactionsBaseOffset, p.localTransactionsModulator
+}
+
+// update the peer once the message was sent successfully.
+func (p *Peer) updateMessageSent(txMsg *transactionBlockMessage, selectedTxnIDs []transactions.Txid, timestamp time.Duration, sequenceNumber uint64, messageSize int, filter bloomFilter) {
+	p.recentSentTransactions.addSlice(selectedTxnIDs, sequenceNumber, timestamp)
+	p.lastSentMessageSequenceNumber = sequenceNumber
+	p.lastSentMessageRound = txMsg.Round
+	p.lastSentMessageTimestamp = timestamp
+	p.lastSentMessageSize = messageSize
+	if filter.filter != nil {
+		p.lastSentBloomFilter = filter
+	}
+}
+
+// setLocalRequestParams stores the peer request params.
+func (p *Peer) setLocalRequestParams(offset, modulator uint64) {
+	if modulator > 255 {
+		modulator = 255
+	}
+	p.localTransactionsModulator = byte(modulator)
+	if modulator != 0 {
+		p.localTransactionsBaseOffset = byte(offset % modulator)
+	}
+}
+
+// peers array functions
+
+// incomingPeersOnly scan the input peers array and return a subset of the peers that are incoming peers.
+func incomingPeersOnly(peers []*Peer) (incomingPeers []*Peer) {
+	incomingPeers = make([]*Peer, 0, len(peers))
+	for _, peer := range peers {
+		if !peer.isOutgoing {
+			incomingPeers = append(incomingPeers, peer)
+		}
+	}
+	return
+}
+
+// incoming related functions
+
+func (p *Peer) addIncomingBloomFilter(round basics.Round, incomingFilter bloomFilter, currentRound basics.Round) {
+	bf := incomingBloomFilter{
+		round:  round,
+		filter: incomingFilter,
+	}
+	// scan the current list and find if we can removed entries.
+	firstValidEntry := sort.Search(len(p.recentIncomingBloomFilters), func(i int) bool {
+		return p.recentIncomingBloomFilters[i].round >= currentRound.SubSaturate(1)
+	})
+	if firstValidEntry < len(p.recentIncomingBloomFilters) {
+		// delete some of the old entries.
+		p.recentIncomingBloomFilters = p.recentIncomingBloomFilters[firstValidEntry:]
+	}
+	// reset the counter, since we might need to re-evaluate some of the transaction group with the new bloom filter.
+	p.lastTransactionSelectionTracker.roll(incomingFilter.encodingParams.Offset, incomingFilter.encodingParams.Modulator)
+
+	// scan the existing bloom filter, and ensure we have only one bloom filter for every
+	// set of encoding parameters. this would allow us to accumulate false positive
+	for idx, bloomFltr := range p.recentIncomingBloomFilters {
+		if bloomFltr.filter.encodingParams == incomingFilter.encodingParams {
+			// replace.
+			p.recentIncomingBloomFilters[idx] = bf
+			return
+		}
+	}
+
+	p.recentIncomingBloomFilters = append(p.recentIncomingBloomFilters, bf)
+	if len(p.recentIncomingBloomFilters) > maxIncomingBloomFilterHistory {
+		p.recentIncomingBloomFilters = p.recentIncomingBloomFilters[1:]
+	}
```

Comment:
First drop then add the new? To avoid running out of space and copying the slice?